lib/mayaUsd/utils/mayaNodeObserver.h

```diff
@@ -0,0 +1,121 @@
+//
+// Copyright 2024 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#ifndef MAYAUSD_MAYA_NODE_OBSERVER_H
+#define MAYAUSD_MAYA_NODE_OBSERVER_H
+
+#include <mayaUsd/base/api.h>
+
+#include <maya/MDagPath.h>
+#include <maya/MMessage.h>
+#include <maya/MObject.h>
+#include <maya/MPlug.h>
+#include <maya/MString.h>
+
+#include <set>
+#include <vector>
+
+namespace MAYAUSD_NS_DEF {
+
+//! Observer for a single Maya node that receives notifications when the node is renamed,
+//  reparented, or any of its ancestor is renamed or reparented.
+//
+// It forwards those notifications to listerners.
+class MayaNodeObserver
+{
+public:
+    //! Listener triggered by the observation.
+    struct MAYAUSD_CORE_PUBLIC Listener
+    {
+        //! Called when a given particular node is renamed.
+        virtual void processNodeRenamed(MObject& observedNode, const MString& str);
+        //! Called when a parent is added anywhere in the hierarchy of a given particular node.
+        virtual void processParentAdded(MObject& observedNode, MDagPath& child, MDagPath& parent);
+        //! Called when a plug is dirtied anywhere in the hierarchy of a given particular node.
+        virtual void processPlugDirty(
+            MObject& observedNode,
+            MObject& dirtiedNode,
+            MPlug&   plug,
+            bool     pathChanged);
+    };
+
+    //! Create a Maya node observer. To start observing the node, call startObserving().
+    MAYAUSD_CORE_PUBLIC
```

Comment:
Is there a reason why you are exporting each of the member functions rather than just exporting the whole class (like you did with the struct Listener)?

Reply:
Yes, exporting the whole classes export the std::unordered_map below, which forces the compiler to generate all functions in the template, but some of these functions are not template-generatable because it uses types that are only moveable, so the compilations fails if we export the whole class.

Most other classes in MayaUSD also only export their public interface, not the private functions.

Reply:
Okay thanks for the explanation. Makes sense.